When AudioVideoRendereRemote is in use, initial currentTime is invalidTime

jyavenard opened the pull request:
#### 15aaeaf42309012cc7eae9b645bc330492eab71d
<pre>
When AudioVideoRendereRemote is in use, initial currentTime is invalidTime
<a href="https://bugs.webkit.org/show_bug.cgi?id=300193">https://bugs.webkit.org/show_bug.cgi?id=300193</a>
<a href="https://rdar.apple.com/161985264">rdar://161985264</a>

Reviewed by Youenn Fablet.

We make the RemoteAudioVideoRendererProxyManager send the AudioVideoRenderer
state upon creation. To do so, we have to send the RemoteAudioVideoRendererProxyManager::Create
message only once the AudioVideoRendereRemoteMessageReceiver is setup, so we
move the create to be performed by the AudioVideoRendererRemote constructor
rather than the AudioVideoRendererRemot::create method.

Added internal method to retrieve MediaPlayerPrivate&apos;s currentTime as the
HTMLMediaElement will return 0 if the currentTime is invalid.

Test: media/media-vp8-webm-initialcurrentTime.html
* LayoutTests/media/media-vp8-webm-initialcurrentTime-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-initialcurrentTime.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerCurrentTime const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::privatePlayerCurrentTime):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
(WebKit::RemoteAudioVideoRendererProxyManager::create):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::create):
(WebKit::AudioVideoRendererRemote::AudioVideoRendererRemote):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: <a href="https://commits.webkit.org/301123@main">https://commits.webkit.org/301123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/773c8dc187b7f5c13a2f9a5da92470b0abfa38c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44681 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131861 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/065d8f15-3a5c-40bc-92ca-58d5ea613cda) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45373 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95158 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/73c5d382-65db-40fa-b204-33cf2cffbaf4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127966 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75701 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/147c9133-9015-4c82-8bd3-44ab6dbde142) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29956 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75343 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30186 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134537 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51834 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39631 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103630 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52259 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103404 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26321 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48745 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27033 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48876 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51719 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57512 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51093 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54450 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52784 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8c4a584e2b58e55e1315e21d8830c569c62b31b2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51832

jernoble commented:
It looks like the mac-wk2 bot showed at least one flake ("NaN == 0") result during the stress testing. Do we need to delay "loadedmetadata" until we get a duration?

jyavenard committed:
When AudioVideoRendereRemote is in use, initial currentTime is invalidTime
https://bugs.webkit.org/show_bug.cgi?id=300193
rdar://161985264

Reviewed by Youenn Fablet.

We make the RemoteAudioVideoRendererProxyManager send the AudioVideoRenderer
state upon creation. To do so, we have to send the RemoteAudioVideoRendererProxyManager::Create
message only once the AudioVideoRendereRemoteMessageReceiver is setup, so we
move the create to be performed by the AudioVideoRendererRemote constructor
rather than the AudioVideoRendererRemot::create method.

Added internal method to retrieve MediaPlayerPrivate's currentTime as the
HTMLMediaElement will return 0 if the currentTime is invalid.

Test: media/media-vp8-webm-initialcurrentTime.html
* LayoutTests/media/media-vp8-webm-initialcurrentTime-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-initialcurrentTime.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerCurrentTime const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::privatePlayerCurrentTime):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
(WebKit::RemoteAudioVideoRendererProxyManager::create):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::create):
(WebKit::AudioVideoRendererRemote::AudioVideoRendererRemote):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: https://commits.webkit.org/301123@main
